core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocatorService.java

```diff
@@ -0,0 +1,274 @@
+        {
+            try {
+                Optional<InternalNode> node = tryAcquireNode(requirements);
+                if (node.isPresent()) {
+                    return immediateFuture(node.get());
+                }
+            }
+            catch (RuntimeException e) {
+                return immediateFailedFuture(e);
+            }
+
+            SettableFuture<InternalNode> future = SettableFuture.create();
+            PendingAcquire pendingAcquire = new PendingAcquire(requirements, future);
+            pendingAcquires.add(pendingAcquire);
+
+            return future;
+        }
+
+        @Override
+        public void release(InternalNode node)
+        {
+            releaseNodeInternal(node);
+            processPendingAcquires();
+        }
+
+        public void updateNodes()
+        {
+            processPendingAcquires();
+        }
+
+        private synchronized Optional<InternalNode> tryAcquireNode(NodeRequirements requirements)
+        {
+            NodeSelector nodeSelector = nodeSelectorCache.computeIfAbsent(requirements.getCatalogName(), catalogName -> nodeScheduler.createNodeSelector(session, catalogName));
+
+            List<InternalNode> nodes = nodeSelector.allNodes();
+            if (nodes.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            List<InternalNode> nodesMatchingRequirements = nodes.stream()
+                    .filter(node -> requirements.getAddresses().isEmpty() || requirements.getAddresses().contains(node.getHostAndPort()))
+                    .collect(toImmutableList());
+
+            if (nodesMatchingRequirements.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            Optional<InternalNode> selectedNode = nodesMatchingRequirements.stream()
+                    .filter(node -> allocationCountMap.getOrDefault(node, 0) < maximumAllocationsPerNode)
+                    .min(comparing(node -> allocationCountMap.getOrDefault(node, 0)));
+
+            if (selectedNode.isEmpty()) {
+                return Optional.empty();
+            }
+
+            allocationCountMap.compute(selectedNode.get(), (key, value) -> value == null ? 1 : value + 1);
+            return selectedNode;
+        }
+
+        private synchronized void releaseNodeInternal(InternalNode node)
+        {
+            int allocationCount = allocationCountMap.compute(node, (key, value) -> value == null ? 0 : value - 1);
+            checkState(allocationCount >= 0, "allocation count for node %s is expected to be greater than or equal to zero: %s", node, allocationCount);
+        }
+
+        private void processPendingAcquires()
+        {
+            verify(!Thread.holdsLock(this));
+
+            IdentityHashMap<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();
```

Comment:
`IdentityHashMap<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();`
->
`Map<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();`

Reply:
This is on purpose. `IdentityHashMap` has a different contract than `Map`. It is more readable to be explicit.